texlive.withPackages: replace postBuild with install-tl like script

xworld21 opened the pull request:
## Description of changes
I have refactored/rewritten the cumbersome `postBuild` script used in `texlive.withPackages`. It now follows very closely the upstream install-tl script. I am even using the same function names and the same log messages.

Summary:
- the script can be shellchecked and easily reviewed against the upstream install-tl
- ConTeXt caches are generated correctly
- custom packages can specify `passthru.fontMaps`, `passthru.formats`, `passthru.hyphenPatterns` – this is because we generate the config files directly instead of using `sed` (this is necessary for integrating tlcontrib, and has been implicitly requested by a user)
- logging and debugging is massively better
- derivation is easier to override
- we are getting closer to `strictDeps` working

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/new-tex-live-withpackages-function/34902/12

xworld21 commented:
@veprbl I think this refactor is ready now (I'd rather do the next texlive improvements, like deduplicating formats, after this is merged – reviews will be easier)

veprbl commented on pkgs/tools/typesetting/tex/texlive/build-tex-env.sh:
```diff
@@ -0,0 +1,327 @@
+TEXMFSYSVAR="$out/share/texmf-var"
+export TEXMFCNF="$TEXMFSYSCONFIG/web2c"
+
+do_postinst_stuff
```
Perhaps convert this to a shellHook `postBuildHooks+=(do_postinst_stuff)` (possibly give it a more texlive-related name)

xworld21 replied:
> Perhaps convert this to a shellHook `postBuildHooks+=(do_postinst_stuff)` (possibly give it a more texlive-related name)

Ah, that's not working... `buildEnv` is based on `runCommand`, which sets `buildCommand`, which bypasses phases and hooks altogether.

I can prefix all function names with `tl_`, though, for future proofing?

xworld21 replied:
(To clarify: `do_postinst_stuff` is literally the name of the corresponding function in `install-tl`, I didn't make it up.)

xworld21 replied:
This made me think that we could use `mkDerivation` directly, and group those functions under traditional phases (configure for `fmtutil.cnf`, build for the formats, install or fixup for the wrappers). We could then switch to the `finalAttrs: prevAttrs: ...` pattern and drop `.overrideTeXConfig`.

Something for 24.11, maybe. I'd rather make sure the current refactor really works, do other minor things (I have tlcontrib ready!), then start experimenting with the 'overrideable texlive' idea right after 24.05.

veprbl commented:
@GrahamcOfBorg build tests.texlive

xworld21 committed:
texlive.tlpdb.nix: add hyphenation patterns and font maps to tlpdb

xworld21 committed:
texlive.withPackages: replace postBuild with install-tl like script

xworld21 committed:
tests.texlive.fmtutilCnf: remove

The fmtutilCnf test does not fit the new way of generating fmtutil.cnf.

xworld21 committed:
tests.texlive.context: check that the PDF has been created

xworld21 commented:
Right, local variable snafu fixed. I have added an improvement to the ConTeXt test and renamed some variables to make the code (marginally) clearer (which made me notice another bug!). Tests succeed on my machine, I checked this time.

@GrahamcOfBorg build tests.texlive

veprbl commented:
@GrahamcOfBorg build tests.texlive
